cmd/stunstamp/stunstamp.go

```diff
@@ -0,0 +1,664 @@
+		r := result{}
+		if conn == nil {
+			var err error
+			if source == timestampSourceKernel {
+				conn, err = getConnKernelTimestamp()
+			} else {
+				conn, err = net.ListenUDP("udp", &net.UDPAddr{})
+			}
+			if err != nil {
+				select {
+				case <-doneCh:
+					return
+				case errCh <- err:
+					return
+				}
+			}
+			defer conn.Close()
+		} else {
+			r.stableConn = true
+		}
+		fn := measureRTT
+		if source == timestampSourceKernel {
+			fn = measureRTTKernel
+		}
+		rtt, err := probe(meta, conn, fn)
+		if err != nil {
+			select {
+			case <-doneCh:
+				return
+			case errCh <- err:
+				return
+			}
+		}
+		r.at = at
+		r.meta = meta
+		r.timestampSource = source
+		r.rtt = rtt
+		select {
+		case <-doneCh:
+		case resultsCh <- r:
+		}
+	}
+
+	for regionID, region := range dm.Regions {
+		for _, node := range region.Nodes {
+			v4, err := netip.ParseAddr(node.IPv4)
+			if err != nil || !v4.Is4() {
+				log.Printf("invalid ipv4 addr for node, skipping: %v", node.Name)
+				continue
+			}
+			v6, err := netip.ParseAddr(node.IPv6)
+			if err != nil || !v6.Is6() {
+				log.Printf("invalid ipv6 addr for node, skipping: %v", node.Name)
+				continue
+			}
+			addrsToProbe[v4] = true
+			addrsToProbe[v6] = true
+			meta4 := nodeMeta{
+				regionID:   regionID,
+				regionCode: region.RegionCode,
+				hostname:   node.HostName,
+				addr:       v4,
+			}
+			meta6 := meta4
+			meta6.addr = v6
+
+			var v6StableConns [2]io.ReadWriteCloser
+			v4StableConns, err := getStableConns(stableConns, meta4.addr)
+			if err != nil {
+				close(doneCh)
+				wg.Wait()
+				return nil, err
+			}
+			if *flagIPv6 {
+				v6StableConns, err = getStableConns(stableConns, meta6.addr)
+				if err != nil {
+					close(doneCh)
+					wg.Wait()
+					return nil, err
+				}
+			}
+
+			wg.Add(1)
+			numProbes += 2
+			go doProbe(v4StableConns[timestampSourceUserspace], meta4, timestampSourceUserspace)
+			go doProbe(nil, meta4, timestampSourceUserspace)
+			if *flagIPv6 {
+				wg.Add(1)
+				numProbes += 2
+				go doProbe(v6StableConns[timestampSourceUserspace], meta6, timestampSourceUserspace)
+				go doProbe(nil, meta6, timestampSourceUserspace)
+			}
+			if runtime.GOOS == "linux" {
+				wg.Add(1)
+				numProbes += 2
+				go doProbe(v4StableConns[timestampSourceKernel], meta4, timestampSourceKernel)
+				go doProbe(nil, meta4, timestampSourceKernel)
+				if *flagIPv6 {
+					wg.Add(1)
+					numProbes += 2
+					go doProbe(v6StableConns[timestampSourceKernel], meta6, timestampSourceKernel)
+					go doProbe(nil, meta6, timestampSourceKernel)
+				}
+			}
+		}
+	}
+
+	// cleanup conns we no longer need
+	for k, conns := range stableConns {
+		if !addrsToProbe[k] {
+			if conns[timestampSourceKernel] != nil {
+				conns[timestampSourceKernel].Close()
+			}
+			conns[timestampSourceUserspace].Close()
+			delete(stableConns, k)
+		}
+	}
+
+	for {
+		select {
+		case err := <-errCh:
+			close(doneCh)
+			wg.Wait()
```

Comment:
I've not found any `wg.Done()` calls on this, is it expected?

Reply:
updated in https://github.com/tailscale/tailscale/pull/12241/commits/41c045be5ffcb84ebe6b3d0794baff17689236e2